File: app/src/main/java/app/proyecto/Activities/PagoActivity.java
package app.proyecto.Activities;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.ListView;
import android.widget.RadioButton;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.EditText;
import android.widget.CheckBox;

import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;

import org.w3c.dom.Text;

import java.util.ArrayList;
import java.util.Collection;

import app.proyecto.Adapters.AdapterCheckBox;
import app.proyecto.R;
import app.proyecto.databinding.ActivityMainBinding;
import app.proyecto.databinding.ActivityPagoBinding;

public class PagoActivity extends AppCompatActivity {

    private FirebaseFirestore db = FirebaseFirestore.getInstance();
    private CollectionReference MesaRef = db.collection( "user"); //Cambiar Valores cuando Rulin haga el método
    private TextView textViewData;
    private EditText Total;
    private RadioButton rb_PayPal, rb_Efectivo;
    private CheckBox Todos;
    private int listo = 0;
    boolean Clientes[];
    int clientes = 0;
    Double i_Total;
    private int i=0;

    private ListView L_ListaClientes;
    private ArrayList<String> S_ListaClientes = new ArrayList<>();
    private ArrayList<Double> T_ListaClientes = new ArrayList<>();
    private AdapterCheckBox adapter = new AdapterCheckBox(S_ListaClientes,this);
    private ActivityPagoBinding PagoGrafico;
    private LinearLayoutManager layoutManager;

    public PagoActivity() {
    }


    @RequiresApi(api = Build.VERSION_CODES.N)
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_pago);

        PagoGrafico = ActivityPagoBinding.inflate(getLayoutInflater());
        PagoGrafico.LListaClientes.setLayoutManager(new LinearLayoutManager(this)); // :O
        PagoGrafico.LListaClientes.setAdapter(adapter);

        textViewData = findViewById(R.id.txt_Total);

        rb_PayPal = (RadioButton) findViewById(R.id.rB_PayPal);
        rb_Efectivo = (RadioButton) findViewById(R.id.rB_Efectivo);
        Todos = (CheckBox) findViewById(R.id.Ch_Todos);
        Total = (EditText) findViewById(R.id.txt_Total);

        L_ListaClientes = (ListView) findViewById(R.id.L_ListaClientes);
        L_ListaClientes.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE);

        MesaRef.get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
            @Override
            public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
                for (QueryDocumentSnapshot documentSnapshot : queryDocumentSnapshots) {
                    S_ListaClientes.add((String) documentSnapshot.get("Nombre"));
                    T_ListaClientes.add((Double) documentSnapshot.get("Total"));
                    i = i+1;
                }
                adapter.notifyDataSetChanged();
            }
        });


        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        Clientes = new boolean[i];
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        finish();
        return super.onOptionsItemSelected(item);
    }

    private void ActualizaClientes (View view){

    }

    private void CambiaTotal (View view) {
        i_Total = 0.0;
        if (L_ListaClientes.getSelectedItem() == null){
            Total.setText("");
        }else{
//            Clientes = L_ListaClientes.getCheckedItemPositions(); //DUDA AAAAAAAAAAAAAAAAAA Guardar
            for (int j=0; j<i; j++){
                if (Clientes[j])
                    i_Total = i_Total + T_ListaClientes.get(j);
            }
            Total.setText("" + i_Total);
        }
    }

    public void SeleccionarTodos (View view){
        int j;
        if (Todos.isChecked()) {
            for (j = 0; j<i; j++)
                L_ListaClientes.setItemChecked(j, true);
            clientes = i;
        }else{
            for (j = 0; j<i; j++)
                L_ListaClientes.setItemChecked(j, false);
            clientes = 0;
        }
        CambiaTotal(view);
    }

    public void Confirmar (View view){

        if (clientes > 0) {
            if (rb_PayPal.isChecked() == true) {
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                Toast.makeText(this, "Pago Realizado con éxito", Toast.LENGTH_LONG).show();
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                //startActivity(oPantallaDespedida);
            } else if (rb_Efectivo.isChecked() == true) {
                try {
                    Thread.sleep(20000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                Toast.makeText(this, "Pago confirmado", Toast.LENGTH_LONG).show();
                try {
                    Thread.sleep(5000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                //startActivity(oPantallaDespedida);
            }else{
                Toast.makeText(this, "Seleccione su forma de pago", Toast.LENGTH_SHORT).show();
            }
        }
    }

}

File: app/src/main/java/app/proyecto/Activities/PagoActivity.kt
package app.proyecto.Activities

import android.content.ContentValues
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.CheckBox
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.GridLayoutManager
import app.proyecto.Adapters.AdapterCheckBox
import app.proyecto.DataAccess.DBUser
import app.proyecto.Models.Integrante
import app.proyecto.R
import app.proyecto.Utils.BusinessLogic
import com.google.common.collect.ArrayListMultimap
import com.google.firebase.firestore.FirebaseFirestore
import kotlinx.android.synthetic.main.activity_pago.*

class PagoActivity : AppCompatActivity() {
    lateinit var restaurant:String
    lateinit var logiB:BusinessLogic
    lateinit var mesa:String
    var Total:Double=0.0
    private val lista:ArrayList<Integrante> = ArrayList();
    lateinit var adap:AdapterCheckBox
    val db : FirebaseFirestore = FirebaseFirestore.getInstance()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_pago)
        logiB=BusinessLogic(this)
        adap= AdapterCheckBox(lista,ChPago_Todos,txtPago_Total,this)
        RVPago_ListaClientes.layoutManager=GridLayoutManager(this,1)
        RVPago_ListaClientes.adapter=adap

        restaurant = intent.getStringArrayExtra("Code")?.get(1).toString()
        mesa = intent.getStringArrayExtra("Code")?.get(2).toString()
        setup()
    }

    private fun setup() {
        llenarLista()

        ChPago_Todos.setOnClickListener(){
            if(ChPago_Todos.isChecked)
                marcarTodos(1)
            else
                marcarTodos(2)
        }

        btnPago_Confirmar.setOnClickListener(){
            if (!rBPago_Efectivo.isChecked && !rBPago_PayPal.isChecked)
                Toast.makeText(this,"Selecciona algun metodo de pago", Toast.LENGTH_SHORT).show()
            else{
                for (integrante in lista){
                    if(integrante.chek.isChecked){
                        db.collection("Restaurants")
                            .document(restaurant)
                            .collection(mesa).
                            document(integrante.nombre).delete().addOnCompleteListener(){
                                if(it.isSuccessful){
                                    DBUser.i--
                                    lista.remove(integrante)
                                    adap.notifyDataSetChanged()
                                    Buscar()
                                }
                            }
                    }
                }
                marcarTodos(2)
                DBUser.TotalGeneral=0.0
                txtPago_Total.text="$ "+DBUser.TotalGeneral
            }

        }

    }

    private fun Buscar(){
        var bandera=false
        for (elem in lista){
            if(elem.nombre.equals(DBUser.miUsuario.Nombre)) {
                bandera = true
                break
            }
        }
        if (!bandera) {
            logiB.deleteLists()
            val prefs = getSharedPreferences(getString(R.string.prefs_codigo), Context.MODE_PRIVATE).edit()
            prefs.clear()
            prefs.apply()
        }
    }

    private fun llenarLista(){
    db.collection("Restaurants")
        .document(restaurant)
                .collection(mesa)
                .get()
                .addOnCompleteListener(){
                    if(it.isSuccessful) {
                        lista.clear()
                        for (documentos in it.result!!) {
                            val nombre = documentos.getString("Nombre")
                            val totalp:Double?= documentos.get("Total") as Double?

                            if (nombre != null &&totalp != null ) {
                                lista.add( Integrante(nombre,totalp) )

                            }
                        }
                        adap.notifyDataSetChanged()
                    }
                    else
                        Log.w(ContentValues.TAG, "No pude leer nada")
        }

            db.collection("Restaurants")
                    .document(restaurant)
                    .collection(mesa).addSnapshotListener{snapshot,exception ->
                        if(exception != null){
                            Toast.makeText(this,"ah sucedido un error",Toast.LENGTH_SHORT).show()
                        }
                        lista.clear();
                        snapshot?.forEach(){
                            val nombre = it.getString("Nombre")
                            val totalp:Double?= it.get("Total") as Double?

                            if (nombre != null &&totalp != null ) {
                                lista.add( Integrante(nombre,totalp) )

                            }

                        }
                        adap.notifyDataSetChanged()
                        Buscar()
                        if(lista.isEmpty())
                            showAlert()
                    }
    }

    private fun marcarTodos(i:Int){
        var valor:Boolean
        valor = i==1

        for (it in lista){
            it.chek.isChecked=valor
            it.chek.isSelected=valor
            if (valor)
                Total+=it.totalParcial
            else
                Total-=it.totalParcial
            }
        DBUser.TotalGeneral=Total
        txtPago_Total.text="$ "+DBUser.TotalGeneral
    }

    private fun showAlert(){
        val builder= AlertDialog.Builder(this)
        builder.setTitle("Mesa Cerrada")
        builder.setMessage("Gracias por su visita, vuelva pronto")
        builder.setPositiveButton("Aceptar", DialogInterface.OnClickListener() { dialogInterface: DialogInterface, i: Int ->
            startActivity(Intent(this, PromocionesActivity::class.java))
            finishAffinity()
            finish()

        })
        val dialog: AlertDialog = builder.create()
        dialog.show()
    }
}

